# Linear Algebra 2 - Dot products and Hyperplanes

### Topics

1. Vectors
2. Norms
3. Matrix Multiplication
4. Dot product
5. Angle between two vectors
6. Unit Vectors
7. Projection of a Vector

## 1 Vector Algebra

### 1.1 Vectors

#### Definition

A quantifiable entity that has **both magnitude and direction** is called as Vector.

#### Properties

1. A Vector has only **one dimension**.
2. Mathematically, a vector $\vec{x}$ is represented as an array of numbers (a column).
3. In ML Context, each data-point (a single row in the dataset) is represented as a vector on a Cartesian plane.

#### Vector Representation

A vector can be represented as a row vector or a column vector, but when it is not explicitly specified, a vector is assumed to be a column vector.

Mathematical notation of a vector $\vec{x}$ having $d$ dimension containing real numbers:

$$
\begin{aligned}
\large
x \in \mathbb{R}^d
\end{aligned}
$$

Where $d$ represents number of elements in the vector.

##### Row Vector

$$
\begin{aligned}
\large
\vec{x} = [x_1 \; x_2 \; \ldots \; x_d]
\end{aligned}
$$

##### Column Vector

$$
\begin{aligned}
\large
\vec{x} = \begin{bmatrix} x_1 \\ x_2 \\ \vdots \\ x_d \end{bmatrix}
\end{aligned}
$$

### 1.2 Magnitude

#### Definition

1. Magnitude is the distance of a data-point from the origin represented as a vector.
2. Magnitude of a vector is the shortest distance (displacement) between the origin and the data-point.

#### Properties

1. Shortest distance between two points is called as Displacement.
2. Displacement between two points is calculated using **Euclidean Distance**.
3. Magnitude of a vector is also called as **Norm** of a vector.
4. Mathematically, Norm of a vector is represented as $\Vert\bar{x}\Vert$.
5. In ML context, Norm of a vector is its length from the origin $(0, 0)$ on a Cartesian plane.

There are two types of Norms:

1. L1 Norm
2. L2 Norm

### 1.3 L2 Norm or Euclidean Distance

#### Formula

Formula for L2 Norm or Euclidean Distance of a vector of $d$ dimension:

$$
\begin{aligned}
\large
{\Vert\vec{x}\Vert}_2 = \sqrt{{x_1}^2 + {x_2}^2 + {x_3}^2 + \ldots + {x_d}^2}
\end{aligned}
$$

or,

$$
\begin{aligned}
\large
{\Vert{x}\Vert}_2 = \sqrt{\sum_{i=1}^d {x_i}^2}
\end{aligned}
$$

### 1.4 L1 Norm or Manhattan Distance

#### Formula

Formula for L1 Norm or Manhattan Distance of a vector of $d$ dimension:

$$
\begin{aligned}
\large
{\Vert\vec{x}\Vert}_1 = \vert{x_1}\vert + \vert{x_2}\vert + \vert{x_3}\vert + \ldots + \vert{x_d}\vert
\end{aligned}
$$

or,

$$
\begin{aligned}
\large
{\Vert{x}\Vert}_1 = \sum_{i=1}^d \vert{x_i}\vert
\end{aligned}
$$

### 1.4 Direction

Direction of a Vector is determined by its angle to the x-axis.

## 2 Matrix Multiplication

### 2.1 Dot Product

#### Properties

1. Dot product calculates the **linear similarity** between two vectors.
2. Higher the value of dot product higher is the similarity.
3. Lower the value of dot product lower is the similarity.

#### Formula

$$
\begin{aligned}
\large
\vec{a} \cdot \vec{b} = a^T \cdot b
\end{aligned}
$$

or,

$$
\begin{aligned}
\large
\vec{a} \cdot \vec{b} = \sum_{i=1}^d x_i y_i
\end{aligned}
$$

#### Dot Product rule

Lets say we have two matrices $A: m \times n$ and $B: n \times k$,

$$
\begin{aligned}
\large
m \times n \Leftrightarrow n \times k = m \times k
\end{aligned}
$$

1. Number of columns in the first matrix should be equal to the number of rows in the second matrix.
2. Matrix multiplication is not commutative, i.e., $A \times B \ne B \times A$

https://www.geogebra.org/m/ETHXK756

### 2.2 Examples

In [1]:
import numpy as np

#### Example #1: Two Vectors

In [2]:
# Vector Dot Product
v1 = np.array([1, 2, 3])
v2 = np.array([4, 5, 6])

In [3]:
[1 * 4 + 2 * 5 + 3 * 6]

[32]

or

In [4]:
v1 @ v2

np.int64(32)

or

In [5]:
np.dot(v1, v2)

np.int64(32)

#### Example #2: Matrix Multiplication

##### Case #1: Same shape

In [6]:
m1 = np.array(
    [
        [1, 2],
        [3, 4],
    ]
)  # 2 X 2
m2 = np.array(
    [
        [5, 6],
        [7, 8],
    ]
)  # 2 X 2

In [7]:
[
    [(1 * 5 + 2 * 7), (1 * 6 + 2 * 8)],
    [(3 * 5 + 4 * 7), (3 * 6 + 4 * 8)],
]  # 2 X 2

[[19, 22], [43, 50]]

or

In [8]:
m1 @ m2

array([[19, 22],
       [43, 50]])

or

In [9]:
np.dot(m1, m2)

array([[19, 22],
       [43, 50]])

##### Case #2: Different shape

In [10]:
m1 = np.array(
    [
        [1, 2, 3],
        [4, 5, 6],
    ]
)  # 2 X 3
m2 = np.array(
    [
        [1, 2, 3, 4],
        [5, 6, 7, 8],
        [9, 0, 1, 2],
    ]
)  # 3 X 4

In [11]:
[
    [(1 * 1 + 2 * 5 + 3 * 9), (1 * 2 + 2 * 6 + 3 * 0), (1 * 3 + 2 * 7 + 3 * 1), (1 * 4 + 2 * 8 + 3 * 2)],
    [(4 * 1 + 5 * 5 + 6 * 9), (4 * 2 + 5 * 6 + 6 * 0), (4 * 3 + 5 * 7 + 6 * 1), (4 * 4 + 5 * 8 + 6 * 2)],
]  # 2 X 4

[[38, 14, 20, 26], [83, 38, 53, 68]]

or

In [12]:
m1 @ m2

array([[38, 14, 20, 26],
       [83, 38, 53, 68]])

or

In [13]:
np.dot(m1, m2)

array([[38, 14, 20, 26],
       [83, 38, 53, 68]])

### 2.3 ML Context

In ML context, the dot product is used to represent the below mapping function:

$$
\begin{aligned}
\large
w_1\;x_1 + w_2\;x_2 + w_0 = 0
\end{aligned}
$$

as,

$$
\begin{aligned}
\large
{w^T}\cdot{x} + w_0 = 0
\end{aligned}
$$

Where ${w^T}\cdot{x}$ is dot product between weight vector $\vec{w}$ and feature vector $\vec{x}$.

#### Example

1. There is a Dataset $X$ with 250 data-points and 4 features i.e., 250 rows and 4 columns.
2. $X$ is of shape: [4, 250] (250 vectors horizontally stacked to form $X$).
3. A weight vector $\vec{w}$ of shape [4, 1] i.e., 4 rows and 1 columns was obtained after optimization.
4. $w^T$ will be of shape: [1, 4]
5. The resulting Decision Boundary (ML model) will be of shape [1, 250] i.e., $1 \times 4 \Leftrightarrow 4 \times 250 = 1 \times 250$
6. ${w^T}\cdot{x} = 1 \times 250$

### 2.4 Usecase

#### Image Similarity

1. Images are converted to vectors (real numbers).
2. Similar images will have higher dot product values.
3. Dissimilar vectors will have lower dot product values.

#### Text Similarity

1. Texts are converted to vectors (real numbers).
2. Similar images will have higher dot product values.
3. Dissimilar vectors will have lower dot product values.

> **Note**:
>
> Computers can process only numbers hence all types of data (text, images, etc.) needs to be converted to numbers before building a ML model.

## 3 Applications of Dot Product

### 3.1 Angle between Vectors

#### Definition

Angle between the vector of the data-point and positive x-axis.

#### Formula

$$
\begin{aligned}
\large
\cos \uptheta = \frac{a^T \cdot b}{\Vert a \Vert \; \Vert b \Vert}
\end{aligned}
$$

In [14]:
angles = [0, 90, 180, 270, 360]

for idx in range(len(angles) - 1):
    a1 = angles[idx] + 1 if angles[idx] else 0
    a2 = angles[idx + 1]
    c1 = np.cos(np.radians(a1)).round(1)
    c2 = np.cos(np.radians(a2)).round(1)

    print(f"[{a1:>3}, {a2:>3}] = {c1:>4}, {c2:>4}")

[  0,  90] =  1.0,  0.0
[ 91, 180] = -0.0, -1.0
[181, 270] = -1.0, -0.0
[271, 360] =  0.0,  1.0


<div style="display: inline-block">

| Quadrant |      Angle | $Cos\;\uptheta$ |
| -------: | ---------: | :-------------- |
|        1 |    0 to 90 | Positive        |
|        2 |  91 to 180 | Negative        |
|        3 | 181 to 270 | Negative        |
|        4 | 271 to 360 | Positive        |

</div>

#### Usecases

Cosine Similarity

### 3.2 Unit Vector

#### Definition

A vector whose magnitude equals to one.

##### Properties

1. Unit vector is used to identify the **direction of a vector**.

#### Formula

$$
\begin{aligned}
\large
\widehat{w} = \frac{\vec{w}}{\Vert{w}\Vert}
\end{aligned}
$$

### 3.3 Projection of a Vector

#### Formula

$$
\begin{aligned}
\large
\text{Projection of $\vec{x}$ on $\vec{y}$} = \vec{x}\cdot\widehat{y}
\end{aligned}
$$

Substituting the formula for $\widehat{y}$ unit vector,

$$
\begin{aligned}
\large
\text{Projection of $\vec{x}$ on $\vec{y}$} = \frac{\vec{x}\cdot\vec{y}}{\Vert{y}\Vert}
\end{aligned}
$$

#### Usecase

1. Projection of a vector is used in Principal Component Analysis (PCA).